include/envoy/network/listener.h

```diff
@@ -114,6 +114,46 @@ class ListenerCallbacks {
   virtual void onNewConnection(ConnectionPtr&& new_connection) PURE;
 };
 
+/**
+ * UDP listener callbacks.
+ */
+class UdpListenerCallbacks {
+public:
+  enum class ErrorCode { SYSCALL_ERROR, UNKNOWN_ERROR };
+
+  virtual ~UdpListenerCallbacks() = default;
+
+  /**
+   * On first packet received on a UDP socket. This allows the callback handler
+   * to establish filter chain (or any other prepararion).
+   *
+   * @param local_address Local bound socket network address.
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onNewConnection(Address::InstanceConstSharedPtr local_address,
+                               Address::InstanceConstSharedPtr peer_address,
+                               Buffer::InstancePtr&& data) PURE;
+  /**
+   * Called whenever data is received by the underlying Udp socket.
+   *
+   * @param local_address Local bound socket network address.
+   * @param peer_address Network address of the peer.
+   * @param data Data buffer received.
+   */
+  virtual void onData(Address::InstanceConstSharedPtr local_address,
+                      Address::InstanceConstSharedPtr peer_address,
+                      Buffer::InstancePtr&& data) PURE;
+
+  /**
+   * Called when there is an error event.
+   *
+   * @param error_code `ErrorCode` for tbe error event.
+   * @param errno System error number.
+   */
+  virtual void onError(const ErrorCode& err_code, int err_no) PURE;
+};
+
```

Comment:
Seems like the listener might also need to be notified when the UDP socket is writable? AIUI, in the TCP case, this interaction is a little hidden: ListenerCallbacks::onNewConnection() provides a Network::Connection, and that Connection is set up with the Dispatcher to receive events when the connection socket is readable or writable: [link](https://github.com/envoyproxy/envoy/blob/f8ba63a4b2a3f0293cfcb44eddcab178f5adc059/source/common/network/connection_impl.cc#L65).  The "is readable" transition is already covered by UdpListenerCallbacks::onData(), but seems like there will need to be something somewhere for the "is writable" side, to let the receipient know when it can flush/write more packets.

Reply:
Good point. So I guess we need another callback `onWriteReady` ?

Reply:
I think it depends on the UdpListenerCallbacks vs. ListenerCallbacks question. If you go with the already existing ListenerCallbacks interface like @alyssawilk is suggesting, then it would make sense to reflect writability inside the Connection implementation, i.e. Event::FileReadyType::Write -> UdpConnectionImpl::onWriteReady() which could flush any buffered packets that are waiting to be written, paired with a UdpConnectionImpl::write() implementation that buffers packets when the socket is not currently writable.

If you go with UdpListenerCallbacks, then yeah, I guess there would need to be something like an onWriteReady(), and it would need to supply something to write to (either the Socket, or an interface for writing a packet).

Reply:
Not sure how to reuse existing `ListenerCallbacks` for UDP since it has connection concepts. I am leaning towards having `onWriteReady` in `UdpListener`. Let me know if you (@alyssawilk) feel strongly about this.

Reply:
This is related to my comment above about whether we are doing connection oriented or not. IMO I'm leaning towads splitting UDP into 2 types: 1) What you have now, which is geared towards datagram, and 2) One listener type which uses normal listener callbacks, and spits out connections which then supports existing filters. Thoughts?